examples/graph/wgan-graphs.py

```diff
@@ -0,0 +1,545 @@
+        self.latent_dim = self.generator.input_shape[-1]
+
+    def compile(self, optimizer_generator, optimizer_discriminator, **kwargs):
+        super(GraphWGAN, self).compile(**kwargs)
+        self.optimizer_generator = optimizer_generator
+        self.optimizer_discriminator = optimizer_discriminator
+        self.metric_generator = keras.metrics.Mean(name="loss_gen")
+        self.metric_discriminator = keras.metrics.Mean(name="loss_dis")
+
+    def train_step(self, inputs):
+
+        if isinstance(inputs[0], tuple):
+            inputs = inputs[0]
+
+        G_real = inputs
+
+        self.batch_size = tf.shape(inputs[0])[0]
+
+        # Train the discriminator for one or more steps
+        for _ in range(self.discriminator_steps):
+            z = tf.random.normal((self.batch_size, self.latent_dim))
+
+            with tf.GradientTape() as tape:
+                G_gen = self.generator(z, training=True)
+                loss = self._loss_discriminator(G_real, G_gen)
+
+            grads = tape.gradient(loss, self.discriminator.trainable_weights)
+            self.optimizer_discriminator.apply_gradients(
+                zip(grads, self.discriminator.trainable_weights)
+            )
+            self.metric_discriminator.update_state(loss)
+
+        # Train the generator for one or more steps
+        for _ in range(self.generator_steps):
+            z = tf.random.normal((self.batch_size, self.latent_dim))
+
+        with tf.GradientTape() as tape:
+            G_gen = self.generator(z, training=True)
+            loss = self._loss_generator(G_gen)
+
+            grads = tape.gradient(loss, self.generator.trainable_weights)
+            self.optimizer_generator.apply_gradients(
+                zip(grads, self.generator.trainable_weights)
+            )
+            self.metric_generator.update_state(loss)
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def _loss_discriminator(self, G_real, G_gen):
+        logits_real = self.discriminator(G_real, training=True)
+        logits_gen = self.discriminator(G_gen, training=True)
+        loss = tf.reduce_mean(logits_gen) - tf.reduce_mean(logits_real)
+        return loss + self._gradient_penalty(G_real, G_gen) * self.gp_weight
+
+    def _loss_generator(self, G_gen):
+        logits_gen = self.discriminator(G_gen, training=True)
+        return -tf.reduce_mean(logits_gen)
+
+    def _gradient_penalty(self, G_real, G_gen):
+        # Unpack graphs
+        A_real, H_real = G_real
+        A_gen, H_gen = G_gen
+
+        # Generate interpolated graphs (A_interp and H_interp)
+        alpha = tf.random.uniform([self.batch_size])
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1, 1))
+        A_interp = (A_real * alpha) + (1 - alpha) * A_gen
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1))
+        H_interp = (H_real * alpha) + (1 - alpha) * H_gen
+
+        # Compute the logits of interpolated graphs
+        with tf.GradientTape() as tape:
```

Comment:
This gradient tape appears to be nested, in this intended?

Reply:
As far as I understand: Yes :)